Source/Dafny/Resolver/BoundsDiscovery.cs

```diff
@@ -0,0 +1,592 @@
+              bounds.Add(new ComprehensionExpr.MapBoundedPool(e1, e0.Type, e1.Type.AsMapType.Arg, e1.Type.AsMapType.Finite));
+            }
+            break;
+          case BinaryExpr.ResolvedOpcode.EqCommon:
+          case BinaryExpr.ResolvedOpcode.SetEq:
+          case BinaryExpr.ResolvedOpcode.SeqEq:
+          case BinaryExpr.ResolvedOpcode.MultiSetEq:
+          case BinaryExpr.ResolvedOpcode.MapEq:
+            var otherOperand = whereIsBv == 0 ? e1 : e0;
+            bounds.Add(new ComprehensionExpr.ExactBoundedPool(otherOperand));
+            break;
+          case BinaryExpr.ResolvedOpcode.Gt:
+          case BinaryExpr.ResolvedOpcode.Ge:
+            Contract.Assert(false);
+            throw new cce.UnreachableException(); // promised by postconditions of NormalizedConjunct
+          case BinaryExpr.ResolvedOpcode.Lt:
+            if (e0.Type.IsNumericBased(Type.NumericPersuasion.Int)) {
+              if (whereIsBv == 0) {
+                // bv < E
+                bounds.Add(new ComprehensionExpr.IntBoundedPool(null, e1));
+              } else {
+                // E < bv
+                bounds.Add(new ComprehensionExpr.IntBoundedPool(Expression.CreateIncrement(e0, 1), null));
+              }
+            }
+            break;
+          case BinaryExpr.ResolvedOpcode.Le:
+            if (e0.Type.IsNumericBased(Type.NumericPersuasion.Int)) {
+              if (whereIsBv == 0) {
+                // bv <= E
+                bounds.Add(new ComprehensionExpr.IntBoundedPool(null, Expression.CreateIncrement(e1, 1)));
+              } else {
+                // E <= bv
+                bounds.Add(new ComprehensionExpr.IntBoundedPool(e0, null));
+              }
+            }
+            break;
+          case BinaryExpr.ResolvedOpcode.RankLt:
+            if (whereIsBv == 0) {
+              bounds.Add(new ComprehensionExpr.DatatypeInclusionBoundedPool(e0.Type.IsIndDatatype));
+            }
+            break;
+          case BinaryExpr.ResolvedOpcode.RankGt:
+            if (whereIsBv == 1) {
+              bounds.Add(new ComprehensionExpr.DatatypeInclusionBoundedPool(e1.Type.IsIndDatatype));
+            }
+            break;
+          default:
+            break;
+        }
+      }
+      return bounds;
+    }
+
+    private static void DiscoverBoundsFunctionCallExpr<VT>(FunctionCallExpr fce, VT bv, List<ComprehensionExpr.BoundedPool> bounds)
+      where VT : IVariable {
+      Contract.Requires(fce != null);
+      Contract.Requires(bv != null);
+      Contract.Requires(bounds != null);
+
+      var older = GetOlderParameters(fce.Function);
+      if (older == null) {
+        return;
+      }
+
+      // local worker function:
+      bool DoesOlderApply(Expression e, string formalName) {
+        return e.Resolved is IdentifierExpr ide && ide.Var == (IVariable)bv && older.Contains(formalName);
+      }
+
+      var isOlderArgument = false;
+      if (!fce.Function.IsStatic && DoesOlderApply(fce.Receiver, "this")) {
```

Comment:
Consider changing this into 

```
if (!fce.Function.IsStatic && DoesOlderApply(fce.Receiver, "this")) {
  isOlderArgument = true;
}

if (!isOlderArgument) {
  ..
}
```

To make it more clear that the else branch code being in a branch is only an optimisation.